In [32]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns
%matplotlib inline

In [33]:
df = pd.read_csv('combineDataset.csv')
df.head()

,Unnamed: 0,Question,Answer
0,0,What is Python?,"A high-level, interpreted programming language."
1,1,Who created Python?,Guido van Rossum.
2,2,When was Python first released?,1991
3,3,Is Python free?,"Yes, it's open-source."
4,4,What does 'Python' mean?,Named after 'Monty Pythonƒ??s Flying Circus'.


In [34]:
df.tail()

,Unnamed: 0,Question,Answer
13973,13973,"how to change [1,2,3,4] to '1234' using python",""""""""""""".join([1, 2, 3, 4])"
13974,13974,Delete every non utf-8 symbols froms string,"line = line.decode('utf-8', 'ignore').encode('..."
13975,13975,How to execute a command in the terminal from ...,os.system(command)
13976,13976,Python MySQL Parameterized Queries,c.execute('SELECT * FROM foo WHERE bar = %s AN...
13977,13977,Convert a string to datetime object in python,"dateobj = datetime.datetime.strptime(datestr, ..."


In [35]:
df.drop(df[['Unnamed: 0']],axis=1,inplace=True)

In [36]:
df.head()

,Question,Answer
0,What is Python?,"A high-level, interpreted programming language."
1,Who created Python?,Guido van Rossum.
2,When was Python first released?,1991
3,Is Python free?,"Yes, it's open-source."
4,What does 'Python' mean?,Named after 'Monty Pythonƒ??s Flying Circus'.


In [37]:
df.tail()

,Question,Answer
13973,"how to change [1,2,3,4] to '1234' using python",""""""""""""".join([1, 2, 3, 4])"
13974,Delete every non utf-8 symbols froms string,"line = line.decode('utf-8', 'ignore').encode('..."
13975,How to execute a command in the terminal from ...,os.system(command)
13976,Python MySQL Parameterized Queries,c.execute('SELECT * FROM foo WHERE bar = %s AN...
13977,Convert a string to datetime object in python,"dateobj = datetime.datetime.strptime(datestr, ..."


In [38]:
df.isnull().sum()

Question      0
Answer      142
dtype: int64

In [39]:
df.dropna(inplace=True,ignore_index=True)

In [40]:
df.isnull().sum()

Question    0
Answer      0
dtype: int64

In [41]:
df.describe()

,Question,Answer
count,13836,13836
unique,13175,13490
top,round number to nearest integer,TRUE
freq,12,36


In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13836 entries, 0 to 13835
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Question  13836 non-null  object
 1   Answer    13836 non-null  object
dtypes: object(2)
memory usage: 216.3+ KB


In [43]:
import re 
import nltk 
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()

In [44]:
corpus = []
for i in range(0,len(df)):
    review = str(df['Question'].iloc[i])
    review = review.lower().split()
    review = [lemmatizer.lemmatize(word) for word in review]
    review = ' '.join(review)
    corpus.append(review)

## Vectorization 

In [45]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()

X = vectorizer.fit_transform(corpus).toarray()
print("Vectorization complete! Shape of X:", X.shape)

Vectorization complete! Shape of X: (13836, 7352)


7352,This is my model's Vocabulary Size. It means that after i cleaned the text (lowercasing, lemmatizing, etc.), the vectorizer found exactly 7,352 unique, meaningful words across all 13,836 questions combined.

In [47]:
import joblib 
import os 

In [49]:
if not os.path.exists("Models"):
    os.makedirs("Models")

In [50]:
joblib.dump(vectorizer,'Models/tfidf_vectorizer.pkl')
joblib.dump(X,'Models/matrix.pkl')
df.to_pickle('Models/cleand_dataframe.pkl')